### PR TITLE
Correlation Id on each request

### PR DESCRIPTION
**Pull request by dor-benatia**
Signed-off-by: Dor Ben-Atia <dorb@activefence.com>

### CODE REVIEW

**Review by rluvaton — COMMENTED** [review]
Great job! left few comments

File: src/code-templates/services/order-service/entry-points/api/routes.ts [hunk 1]
```diff
@@ -1,15 +1,21 @@
 import util from "util";
 import express from "express";
 import * as newOrderUseCase from "../../domain/new-order-use-case";
+import { AsyncLocalStorage } from 'async_hooks';
 import { errorHandler } from "@practica/error-handling";
 
-export const defineRoutes = (expressApp: express.Application) => {
+export const defineRoutes = (expressApp: express.Application, asyncLocalStorage: AsyncLocalStorage<Map<string, string>>) => {
```
Comment (goldbergyoni) [inline src/code-templates/services/order-service/entry-points/api/routes.ts:7]: Since the same functionality, getting the store, is needed in many other modules (e.g. DB access file), we can't pass als (asynclocalstorage) to every function. Consequently, we need to hold a singleton instance in a module so every potential caller can use it

e.g., module.exports = new AsyncLocalStorage

Then every caller who require it will get the singleton

File: src/code-templates/services/order-service/entry-points/api/routes.ts [hunk 2]
```diff
@@ -1,15 +1,21 @@
...
   const router = express.Router();
 
+  function getIdFromRequest() {
+    return asyncLocalStorage.getStore()?.get("practicaRequestId");
+  }
+
   router.post("/", async (req, res, next) => {
     try {
+      const id = getIdFromRequest();
       console.log(
```
[the remark this answers is not in the document]
Comment (goldbergyoni) [reply src/code-templates/services/order-service/entry-points/api/routes.ts:17]: Raz's idea sounds great IMO

File: src/code-templates/services/order-service/entry-points/api/server.ts [hunk 3]
```diff
@@ -7,16 +7,28 @@ import { defineRoutes } from "./routes";
...
 import configurationSchema from "../../config";
+import { AsyncLocalStorage } from 'async_hooks';
+import { v4 } from 'uuid';
 
 let connection: Server;
+const asyncLocalStorage = new AsyncLocalStorage<Map<string, string>>();
+
+function requestId(req, res, next) {
+    asyncLocalStorage.run(new Map(), () => {
+      asyncLocalStorage.getStore()?.set("practicaRequestId", v4());
```
Comment (goldbergyoni) [inline src/code-templates/services/order-service/entry-points/api/server.ts:18]: The ? optional chaining hides a potential failure. If there is no store, something is wrong and we should make it observable by calling errorHandler.handleError(new AppError('cant-initialize-store', ....))

File: src/code-templates/services/order-service/entry-points/api/server.ts [hunk 4]
```diff
@@ -7,16 +7,28 @@ import { defineRoutes } from "./routes";
...
 import * as configurationProvider from "@practica/configuration-provider";
 import configurationSchema from "../../config";
+import { AsyncLocalStorage } from 'async_hooks';
+import { v4 } from 'uuid';
 
 let connection: Server;
+const asyncLocalStorage = new AsyncLocalStorage<Map<string, string>>();
+
+function requestId(req, res, next) {
+    asyncLocalStorage.run(new Map(), () => {
```
[the remark this answers is not in the document]
Comment (goldbergyoni) [reply src/code-templates/services/order-service/entry-points/api/server.ts:17]: Good point.

Going by the docs, a store will get GCed when its associated resource (request), if I get it right - no worries here. However, we do need to set one also per process as those are not GCed unless explicitly requested

"Calling asyncLocalStorage.disable() is required before the asyncLocalStorage can be garbage collected. This does not apply to stores provided by the asyncLocalStorage, as those objects are garbage collected along with the corresponding async resources."

File: src/code-templates/services/order-service/entry-points/api/server.ts (hunk 3 above)
Comment (goldbergyoni) [inline src/code-templates/services/order-service/entry-points/api/server.ts:18]: Eventually this code will be used by various business apps, they probably don't want the name 'practica' in their code. Shall we call this just requestId

File: src/code-templates/services/order-service/entry-points/api/server.ts [hunk 5]
```diff
@@ -7,16 +7,28 @@ import { defineRoutes } from "./routes";
...
 // ️️️✅ Best Practice: API exposes a start/stop function to allow testing control WHEN this should happen
 async function startWebServer(): Promise<AddressInfo> {
   // ️️️✅ Best Practice: Declare a strict configuration schema and fail fast if the configuration is invalid
   configurationProvider.initialize(configurationSchema);
   const expressApp = express();
   expressApp.use(bodyParser.json());
-  defineRoutes(expressApp);
+  expressApp.use(requestId);
+  defineRoutes(expressApp, asyncLocalStorage);
+  defineStatusCheckerHandler(expressApp);
```
Comment (goldbergyoni) [inline src/code-templates/services/order-service/entry-points/api/server.ts:31]: Awesome feature 💪

I would split into a different PR

File: src/code-templates/services/order-service/entry-points/api/routes.ts [hunk 6]
```diff
@@ -1,15 +1,21 @@
 import util from "util";
 import express from "express";
 import * as newOrderUseCase from "../../domain/new-order-use-case";
+import { AsyncLocalStorage } from 'async_hooks';
```
Comment (goldbergyoni) [inline src/code-templates/services/order-service/entry-points/api/routes.ts:4]: Is this the built-in Node's module?

I think it should be:

import { AsyncLocalStorage } from 'node:async_hooks';

File: src/code-templates/services/order-service/package.json [hunk 7]
```diff
@@ -29,13 +29,15 @@
     "@types/sequelize": "^4.28.11",
     "ajv": "^8.11.0",
     "amqplib": "^0.8.0",
+    "async_hooks": "1.0.0",
```
Comment (goldbergyoni) [inline src/code-templates/services/order-service/package.json:32]: No need to use an external package IMO?

File: src/code-templates/services/order-service/entry-points/api/server.ts [hunk 8]
```diff
@@ -7,16 +7,28 @@ import { defineRoutes } from "./routes";
 import { errorHandler } from "@practica/error-handling";
 import * as configurationProvider from "@practica/configuration-provider";
 import configurationSchema from "../../config";
+import { AsyncLocalStorage } from 'async_hooks';
+import { v4 } from 'uuid';
 
 let connection: Server;
+const asyncLocalStorage = new AsyncLocalStorage<Map<string, string>>();
+
+function requestId(req, res, next) {
```
Comment (goldbergyoni) [inline src/code-templates/services/order-service/entry-points/api/server.ts:16]: Since it's a cross Microservice thing, and we also need to put somewhere a global 'new AsyncLocalStorage' - Shall we make this a library?

**Review by goldbergyoni — CHANGES_REQUESTED** [review]
@dor-benatia @zivkaziv This is a super awesome contribution with a super strategic topic 🛰💚

Since it's an infrastructural and big thing, naturally, it demands multiple iterations. This was a great start which put us close to destination, we have few steps to make it 'practica ready'

Practically we need few hours of more work here over the next two weeks. @dor-benatia and Ziv, can you support this? If not, it's 120% cool, you already brought this topic closer to bed and we can also spent few hours here. Your call, just leme know whether you'd like to proceed here

@rluvaton Thanks for the comments

File: src/code-templates/services/order-service/entry-points/api/server.ts (hunk 3 above)
Comment (rluvaton) [inline src/code-templates/services/order-service/entry-points/api/server.ts:18]: Instead of v4 please use `randomUUID` from `crypto` node modules

**Comment by rluvaton** [comment]
Hey, as I understand from your discussion with @goldbergyoni it was ok that I continued, so I'm closing as it is already merged in some other PR

thanks!
